### DISCUSSION TITLE
FFT Producing Unexpected Outputs

### DISCUSSION THREAD

**Discussion by iemcd** [ImageMagick/ImageMagick#5961] [Help]
### ImageMagick version

7.0.10-57

### Operating system, version and so on

Mageia 8

### Description

I am trying to follow the [Fourier transform tutorial](https://www.imagemagick.org/Usage/fourier/), but am having difficulty. The output of
`$ magick SIPI.tiff -fft -ift SIPI_roundrip.png`
appears almost uniformly white.
`$ magick SIPI.tiff -fft +depth + adjoin lena_fft_%d.png`
shows a very bright magnitude image (white, with blue corners), opposite of what I would expect.
`$ magick SIPI_fft_0.png SIPI_fft_1.png -ift SIPI_restored.png`
Looks similar to the phase-only image from later in the tutuorial, although I cannot say if it is exactly. Other test images have similar failures. I do not know where to begin troubleshooting this. My installed version of lib64fftw3 is 3.3.9.

**Comment by fmw42** [comment]
What do you get from

`magick -version`

does it include FFTW

Post your tiff file (zip it and post the zip)

**Reply by iemcd** [reply]
The output of `$ magick --version` is
```
Version: ImageMagick 7.0.10-57 Q16 x86_64 2021-01-12 https://imagemagick.org
Copyright: © 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenMP(4.5) 
Delegates (built-in): bzlib cairo fftw fontconfig freetype gvc heic jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png rsvg tiff webp wmf x xml zlib
```
I have attached the test image, as well as the other generated images.
[SIPI_FFT_bug.zip](https://github.com/ImageMagick/ImageMagick/files/10395670/SIPI_FFT_bug.zip)

It is worth noting that since I got it working at all (previously, fftw had not been listed), my package manager would like me to move to some other version of imagemagick (I think from a "tainted" repo, I do not know if the version number is different). If I do, then fftw is no longer listed as a delegate, and none of the commands listed here will run at all. I can't think of a mechanism by which this is related, but I also don't want it to come out that this was useful information later that I jut never mentioned.

**Comment by fmw42** [comment]
Works fine for me on IM 7.1.0.57 Q16 Mac OSX Monterey with FFTW 3.3.10

[SIPI.tiff.zip](https://github.com/ImageMagick/ImageMagick/files/10395774/SIPI.tiff.zip)


```
magick SIPI.tiff -fft -ift SIPI_rt.tiff
magick compare -metric rmse SIPI.tiff SIPI_rt.tiff null:
0.00146539 (2.23604e-08)

```

[SIPI_rt.tiff.zip](https://github.com/ImageMagick/ImageMagick/files/10395775/SIPI_rt.tiff.zip)


**What is your version of FFTW?**

**Reply by iemcd** [reply]
My fftw version is 3.3.9.

**Reply by fmw42** [reply]
My only other thought would be to disable OpenMP and try again.

Try

`magick -limit thread 1 SIPI.tiff -fft -ift SIPI_rt.tiff`

**Reply by iemcd** [reply]
No joy, the output is identical (visually and by RMSE).

**Reply by fmw42** [reply]
The output should be identical to the input. What do you mean  "identical" to what?

**Reply by iemcd** [reply]
Identical to if I don't limit the threads.

**Comment by fmw42** [comment]
How was FFTW installed?

**Reply by iemcd** [reply]
It was installed by my package manager. Possibly, I think it was included with the OS, as alsa-utils, gstreamer, and several applications depend on it.

**Comment by fmw42** [comment]
Take this image:

![square31](https://user-images.githubusercontent.com/13458218/211923542-30325b66-6244-4fe6-9d08-e59211a1abf3.png)

and do the following:

`magick square31.png -fft square31_fft.png`

![square31_fft-0](https://user-images.githubusercontent.com/13458218/211923635-6ee7d6b2-04db-458a-9a87-be0e9c1a035a.png)

![square31_fft-1](https://user-images.githubusercontent.com/13458218/211923669-f6846764-bbf1-4bf2-8167-608da2ae4860.png)


`magick square31_fft-0.png -evaluate log 100000 square31_fft_spectrum.png`

![square31_fft_spectrum](https://user-images.githubusercontent.com/13458218/211923714-971f4c25-d7d3-4d48-b063-8af4058a8d56.png)


**Do you get similar results?**

**Reply by iemcd** [reply]
No. Roughly, the colors of both images are inverted, and the logarithmic scaled image looks like graph paper (flat black grid on flat white background).
[square31_IEM.zip](https://github.com/ImageMagick/ImageMagick/files/10396188/square31_IEM.zip)

**Reply by iemcd** [reply]
My mistake, I believe the phase image is the same.

**Reply by fmw42** [reply]
Perhaps an install issue. Can you remove your install and try again for both Imagemagick and FFTW? Perhaps some issue with your package manager.  This is now beyond my abilities if a reinstall does not work.  One of the Imagemagick developers would then have to help.

**Reply by iemcd** [reply]
I think I somehow had a mismatched version of imagemagick and the libmagick library. Uninstalling all of it and reinstalling seems to have picked the right libraries. (fftw is too deeply dependent to easily remove, but rpm verifies it.) Thanks for bearing with me through this strange issue.
